src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php

```diff
@@ -107,7 +120,32 @@ public function vote(TokenInterface $token, $object, array $attributes)
      * @param object               $object
      * @param UserInterface|string $user
      *
+     * @deprecated This method will be removed in 3.0 - override voteOnAttribute instead.
+     *
      * @return bool
      */
-    abstract protected function isGranted($attribute, $object, $user = null);
+    protected function isGranted($attribute, $object, $user = null)
+    {
+        return false;
+    }
+
+    /**
+     * Perform a single access check operation on a given attribute, object and (optionally) user
+     * It is safe to assume that $attribute and $object's class pass supportsAttribute/supportsClass
+     * $user can be one of the following:
+     *   a UserInterface object (fully authenticated user)
+     *   a string               (anonymously authenticated user).
+     *
+     * This method will become abstract in 3.0.
+     *
+     * @param string         $attribute
+     * @param object         $object
+     * @param TokenInterface $token
+     *
+     * @return bool
+     */
+    protected function voteOnAttribute($attribute, $object, TokenInterface $token)
+    {
+        return false;
```

Comment:
actually, this could even be simplified a lot, avoiding the Reflection-based logic:

- always call ``voteOnAttribute`` in ``vote``
- throw a deprecation warning in ``AbstractVoter::voteOnAttribute`` saying it should be overwritten
- implement it as ``return $this->isGranted($attribute, $object, $token->getUser());``